### PR TITLE
Fix: Date Localization and Robust Format Detection for WhatsApp Chat Exports

### LINKED ISSUE #69: Changing date conventions
Hi is it possible to change the  date conventions.
currently its MM//DD/YYYY
i would like to change it to DD/MM/YYYY

### PR DESCRIPTION
**Pull request by rodrigogs**
# Fix: Date Localization and Robust Format Detection for WhatsApp Chat Exports

## Problem Statement

**Issue #69**: WhatsApp chat exports fail to parse correctly when:
1. Date formats don't match the hardcoded US format (MM/DD/YYYY, HH:MM AM/PM)
2. User's locale uses different separators, AM/PM variants, or year placement
3. Timestamps display in the user's browser locale on the UI (currently falls back to browser default)

Real-world example from issue reports:
- Export with `2024/12/01, 10:04 pm` (year-first with lowercase AM/PM) → **parser fails entirely**
- Timestamps in UI show inconsistently based on browser, not user's selected language

### 1. **Intelligent Format Detection**
- Sample first 200 lines of chat to detect predominant date format
- Score each pattern independently (allows multiple formats in same file)
- Resolve day/month ambiguity using evidence from actual data (e.g., if any value > 12)
- Graceful fallback to all patterns if no clear winner

### 2. **Robust Line Normalization** (before regex parsing)
- **Remove invisible marks**: Bidirectional control characters (U+200E, U+200F)
- **Standardize spacing**: Unicode non-breaking spaces (U+00A0, U+202F) → ASCII space
- **Normalize AM/PM**: All variants (`am`, `pm`, `a.m.`, `p.m.`, `p. m.`) → uppercase `AM`/`PM`
- **Convert dot time**: Finnish format `HH.MM.SS` → `HH:MM:SS`
- **Remove language markers**: Finnish `klo` token

### 3. **Extended Format Coverage**
Now supports 12+ date/time format variants:

| Format | Example | Locale |
|--------|---------|--------|
| MM/DD/YY, HH:MM AM/PM | 12/25/24, 3:45 PM | US |
| **DD/MM/YY, HH:MM AM/PM** | 25/12/24, 3:45 PM | Europe (NEW) |
| DD/MM/YY, HH:MM | 25/12/24, 15:45 | Europe/Brazil |
| YYYY/MM/DD, HH:MM AM/PM | 2024/12/25, 10:04 pm | Asia (NEW) |
| YYYY/MM/DD, HH:MM | 2024/12/25, 15:45 | Asia |
| DD.MM.YY, HH:MM | 25.12.24, 15:45 | German |
| DD.MM.YY HH.MM.SS: | 25.12.24 15.25.15: | Finnish (NEW) |
| DD-MM-YY, HH:MM | 25-12-24, 15:45 | European variant |
| [DD/MM/YY, HH:MM:SS] | [25/12/24, 15:45:30] | iOS/old |
| [YYYY/MM/DD, HH:MM:SS] | [2024/12/25, 15:45:30] | Year-first (NEW) |
| [DD/MM/YY, HH:MM:SS AM/PM] | [25/12/24, 3:45:30 PM] | iOS 12h |

### 4. **Consistent UI Localization**
- Fixed emoji lightbox missing explicit locale in `MediaGallery.svelte`
- All timestamps now use browser's selected language via `getLocale()`

### Code Changes
- **`src/lib/parser/chat-parser.ts`** (+191 lines)
  - `normalizeLineForParsing()`: Smart line cleaning function
  - `detectDatePatternOrder()`: Improved pattern scoring with tie-breaking
  - New regex patterns for DD/MM AM/PM, dot-time, bracketed year-first
  - Pattern index constants for easier tie-breaking logic

- **`src/lib/components/MediaGallery.svelte`** (+1 line)
  - Fixed timestamp formatting in lightbox: `.toLocaleString()` → `.toLocaleString(getLocale())`

### Test Coverage
Added 4 new parser test files + updated documentation:

- **`european-ampm-format-test.txt`**: Tests `23/06/2018, 1:55 p.m.` variants
- **`asian-ampm-format-test.txt`**: Tests `2024/12/01, 10:04 pm` (reported in issue)
- **`dot-time-klo-format-test.txt`**: Tests `13.6.2018 klo 21.25.15` (Finnish)
- **`bracketed-year-first-test.txt`**: Tests `[2024/06/13, 21:25:15]` variants
- **`examples/parser-tests/README.md`**: Updated to document all 12+ supported formats

## Testing Strategy

1. **Existing tests**: All pass (0 errors, 0 warnings)
   - `npm run lint` ✅ (Biome formatting)
   - `npm run check` ✅ (TypeScript + Svelte)

2. **Manual validation**:
   - Upload example files from `examples/parser-tests/` as ZIP
   - Verify parser detects correct format automatically
   - Confirm timestamps display in selected locale

3. **Edge cases covered**:
   - Mixed formats in same file (detects predominant)
   - Ambiguous day/month values (uses heuristic tie-breaking)
   - Unicode whitespace and invisible characters
   - Multiple AM/PM variants in same file

### ✅ Fixes
- Issue #69: Users can now import WhatsApp exports in ANY supported locale
- Robust parsing: Won't crash on unexpected AM/PM variants or spacing
- Consistent UI: Timestamps always display in user's selected language

### 🎯 Backwards Compatible
- All existing formats still work (no breaking changes)
- Default behavior improved: detects format instead of assuming US

### 📊 Coverage
- From **7 formats** → **12+ formats** supported
- From **hardcoded order** → **smart detection** with fallback

## Notes for Reviewers

- **Line normalization**: Done once per line before regex, minimal performance impact
- **Pattern detection**: Samples first 200 lines (configurable), O(n) scan
- **Tie-breaking logic**: Only applied when 2+ patterns have same frequency (rare)
- **Locale handling**: Uses existing `getLocale()` from Paraglide i18n system

## Related
- Resolves #69
- Improves resilience for global user base across 10+ locales

### CODE REVIEW

File: src/lib/parser/chat-parser.ts [hunk 1]
```diff
@@ -412,17 +587,22 @@ function isSystemMessage(content: string): boolean {
...
-	for (const { regex, parse } of DATE_PATTERNS) {
-		const match = line.match(regex);
+	const normalizedLine = normalizeLineForParsing(line);
+
+	for (const { regex, parse } of datePatterns) {
+		const match = normalizedLine.match(regex);
 		if (match) {
 			const timestamp = parse(match);
-			const remainder = line.substring(match[0].length);
+			const remainder = normalizedLine.substring(match[0].length);
```
[the remark this answers is not in the document]
Comment (rodrigogs) [reply src/lib/parser/chat-parser.ts]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/rodrigogs/whats-reader/pull/70#discussion_r2875349754)


lint and build before finishing
